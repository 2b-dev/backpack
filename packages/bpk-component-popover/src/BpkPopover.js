import PropTypes from 'prop-types';
import React from 'react';
import { BpkButtonLink } from 'bpk-component-link';
import BpkCloseButton from 'bpk-component-close-button';
import { TransitionInitialMount, cssModules } from 'bpk-react-utils';

import STYLES from './bpk-popover.scss';
import { ARROW_ID } from './constants';

const getClassName = cssModules(STYLES);

const BpkPopover = (props) => {
  const {
    id,
    onClose,
    label,
    closeButtonText,
    children,
    className,
    padded,
    labelAsTitle,
    closeButtonIcon,
    ...rest
  } = props;

  const classNames = [getClassName('bpk-popover')];
  const bodyClassNames = [];

  // outer classNames
  if (className) { classNames.push(className); }

  // inner classNames
  if (padded) { bodyClassNames.push(getClassName('bpk-popover__body--padded')); }

  const labelId = `bpk-popover-label-${id}`;

  return (
    <TransitionInitialMount
      appearClassName={getClassName('bpk-popover--appear')}
      appearActiveClassName={getClassName('bpk-popover--appear-active')}
      transitionTimeout={200}
    >
      <section
        id={id}
        tabIndex="-1"
        role="dialog"
        aria-labelledby={labelId}
        className={classNames.join(' ')}
        {...rest}
      >
        <span id={ARROW_ID} className={getClassName('bpk-popover__arrow')} role="presentation" />
        <div className={getClassName('bpk-popover__inner')}>
          {labelAsTitle
            ? (
              <header className={getClassName('bpk-popover__header')}>
                <h2 id={labelId} className={getClassName('bpk-popover__heading')}>{label}</h2>
                &nbsp;
                {closeButtonIcon ?
                  <BpkCloseButton
                    className={getClassName('bpk-popover__close-button')}
                    label={closeButtonText}
                    onClick={(event) => { props.onClose(event, { source: 'CLOSE_BUTTON' }); }}
                  />
                : <BpkButtonLink
                  onClick={(event) => {
                    props.onClose(event, { source: 'CLOSE_LINK' });
                  }}
                >{closeButtonText}</BpkButtonLink>
                }
              </header>
            ) : (
              <span id={labelId} className={getClassName('bpk-popover__label')}>{label}</span>
            )
          }
          <div className={bodyClassNames.join(' ')}>{children}</div>
          {!labelAsTitle && (
            <footer className={getClassName('bpk-popover__footer')}>
              <BpkButtonLink
                onClick={(event) => {
                  props.onClose(event, { source: 'CLOSE_LINK' });
                }}
              >{closeButtonText}</BpkButtonLink>
            </footer>
          )}
        </div>
      </section>
    </TransitionInitialMount>
  );
};

BpkPopover.propTypes = {
  id: PropTypes.string.isRequired,
  onClose: PropTypes.func.isRequired,
  label: PropTypes.string.isRequired,
  closeButtonText: PropTypes.string.isRequired,
  children: PropTypes.node.isRequired,
  className: PropTypes.string,
  padded: PropTypes.bool,
  labelAsTitle: PropTypes.bool,
  closeButtonIcon: PropTypes.bool,
};

BpkPopover.defaultProps = {
  className: null,
  padded: true,
  labelAsTitle: false,
  closeButtonIcon: true,
};

export default BpkPopover;
