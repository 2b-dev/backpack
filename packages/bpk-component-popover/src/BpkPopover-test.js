import React from 'react';
import renderer from 'react-test-renderer';
import BpkPopover from './BpkPopover';

describe('BpkPopover', () => {
  it('should render correctly', () => {
    const tree = renderer.create(
      <BpkPopover
        onClose={() => null}
        closeButtonText="Close"
      >
        My popover content
      </BpkPopover>,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should render correctly with "padded" attribute equal to false', () => {
    const tree = renderer.create(
      <BpkPopover
        onClose={() => null}
        closeButtonText="Close"
        padded={false}
      >
        My popover content
      </BpkPopover>,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should render correctly with "className" attribute', () => {
    const tree = renderer.create(
      <BpkPopover
        onClose={() => null}
        closeButtonText="Close"
        className="my-custom-class"
      >
        My popover content
      </BpkPopover>,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
