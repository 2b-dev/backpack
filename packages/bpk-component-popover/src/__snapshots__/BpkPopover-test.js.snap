exports[`BpkPopover should render correctly 1`] = `
<section
  aria-labelledby="bpk-popover-label-my-popover"
  className="bpk-popover"
  id="my-popover"
  role="dialog"
  tabIndex="-1">
  <span
    className="bpk-popover__arrow"
    role="presentation" />
  <div
    className="bpk-popover__inner">
    <span
      className="bpk-popover__label"
      id="bpk-popover-label-my-popover">
      My popover
    </span>
    <div
      className="bpk-popover__body bpk-popover__body--padded">
      My popover content
    </div>
    <footer
      className="bpk-popover__footer">
      <button
        className="bpk-link"
        onClick={[Function]}
        type="button">
        Close
      </button>
    </footer>
  </div>
</section>
`;

exports[`BpkPopover should render correctly with "className" attribute 1`] = `
<section
  aria-labelledby="bpk-popover-label-my-popover"
  className="bpk-popover my-custom-class"
  id="my-popover"
  role="dialog"
  tabIndex="-1">
  <span
    className="bpk-popover__arrow"
    role="presentation" />
  <div
    className="bpk-popover__inner">
    <span
      className="bpk-popover__label"
      id="bpk-popover-label-my-popover">
      My popover
    </span>
    <div
      className="bpk-popover__body bpk-popover__body--padded">
      My popover content
    </div>
    <footer
      className="bpk-popover__footer">
      <button
        className="bpk-link"
        onClick={[Function]}
        type="button">
        Close
      </button>
    </footer>
  </div>
</section>
`;

exports[`BpkPopover should render correctly with "labelAsTitle" attribute 1`] = `
<section
  aria-labelledby="bpk-popover-label-my-popover"
  className="bpk-popover"
  id="my-popover"
  role="dialog"
  tabIndex="-1">
  <span
    className="bpk-popover__arrow"
    role="presentation" />
  <div
    className="bpk-popover__inner">
    <header
      className="bpk-popover__header">
      <h4
        className="bpk-heading-h4 bpk-heading--no-bottom-margin"
        id="bpk-popover-label-my-popover">
        My popover
      </h4>
      Â 
      <button
        aria-label="Close"
        className="bpk-close-button bpk-popover__close-button"
        onClick={[Function]}
        title="Close"
        type="button">
        <svg
          className="bpk-close-button__icon bpk-icon--align-to-button"
          height="18"
          viewBox="0 0 18 18"
          width="18"
          xmlns="http://www.w3.org/2000/svg">
          <path
            d="M11.1 9l3.4-3.4c.3-.3.3-.8 0-1.1l-1.1-1.1c-.3-.3-.8-.3-1.1 0L9 6.9 5.6 3.5c-.3-.3-.8-.3-1.1 0l-1 1.1c-.3.3-.3.8 0 1.1L6.9 9l-3.4 3.4c-.3.3-.3.8 0 1.1l1.1 1.1c.3.3.8.3 1.1 0L9 11.1l3.4 3.4c.3.3.8.3 1.1 0l1.1-1.1c.3-.3.3-.8 0-1.1L11.1 9z" />
        </svg>
      </button>
    </header>
    <div
      className="bpk-popover__body bpk-popover__body--padded">
      My popover content
    </div>
  </div>
</section>
`;

exports[`BpkPopover should render correctly with "padded" attribute equal to false 1`] = `
<section
  aria-labelledby="bpk-popover-label-my-popover"
  className="bpk-popover"
  id="my-popover"
  role="dialog"
  tabIndex="-1">
  <span
    className="bpk-popover__arrow"
    role="presentation" />
  <div
    className="bpk-popover__inner">
    <span
      className="bpk-popover__label"
      id="bpk-popover-label-my-popover">
      My popover
    </span>
    <div
      className="bpk-popover__body">
      My popover content
    </div>
    <footer
      className="bpk-popover__footer">
      <button
        className="bpk-link"
        onClick={[Function]}
        type="button">
        Close
      </button>
    </footer>
  </div>
</section>
`;
